src/location/util.rs

```diff
@@ -15,51 +15,504 @@
  * #L%
  */
 
-use std::{collections::HashMap, i32, i64, u8};
+use std::{
+    collections::HashMap,
+    i32, i64,
+    sync::{atomic::AtomicUsize, Arc},
+    u8,
+};
 
+use chrono::Utc;
 use mysql as my;
-use mysql::{prelude::Queryable, Row};
+use mysql::{prelude::Queryable, PooledConn, Row};
 
 use super::{
     ob_part_constants, part_func_type::PartFuncType, ObHashPartDesc, ObKeyPartDesc, ObPartDesc,
     ObPartitionInfo, ObPartitionLevel, ObRangePartDesc, TableEntry,
 };
 use crate::{
+    constant::ALL_DUMMY_TABLE,
     error::{CommonErrCode, Error::Common as CommonErr, Result},
+    location::{
+        ObPartitionEntry, ObPartitionLocation, ObReplicaType, ObServerAddr, ObServerInfo,
+        ObServerRole, ObServerStatus, ReplicaLocation, TableEntryKey, TableLocation, OB_INVALID_ID,
+    },
     rpc::protocol::partition::{
         ob_column::{ObColumn, ObGeneratedColumn, ObSimpleColumn},
         ob_partition_key::{Comparable, ObPartitionKey},
     },
     serde_obkv::value::{CollationLevel, CollationType, ObjMeta, ObjType, Value},
+    util::obversion::{ob_vsn_major, parse_ob_vsn_from_sql},
+    ResultCodes,
 };
 
 pub const TEMPLATE_PART_ID: i32 = -1;
 
-//pub const PROXY_PART_INFO_SQL: &str = "SELECT /*+READ_CONSISTENCY(WEAK)*/
-// part_level, part_num, part_type, part_space, part_expr,    part_range_type,
-// part_interval_bin, interval_start_bin,    sub_part_num, sub_part_type,
-// sub_part_space,    sub_part_range_type, def_sub_part_interval_bin,
-// def_sub_interval_start_bin, sub_part_expr,    part_key_name, part_key_type,
-// part_key_idx, part_key_extra, spare1    FROM oceanbase.
-// __all_virtual_proxy_partition_info    WHERE table_id = {} group by
-// part_key_name order by part_key_name LIMIT {};";
-
 pub struct LocationUtil {}
 
 impl LocationUtil {
-    pub fn fetch_partition_info(
+    pub fn get_ob_version_from_server(conn: &mut my::PooledConn) -> Result<()> {
+        if ob_vsn_major() == 0 {
+            let sql = "SELECT /*+READ_CONSISTENCY(WEAK)*/ OB_VERSION() AS CLUSTER_VERSION";
+            for row in conn.query::<Row, &str>(sql)? {
+                let cluster_version = match my::from_row_opt(row) {
+                    Ok(version) => version,
+                    Err(e) => {
+                        error!("ObTableLocation::get_table_entry_from_remote: fail to do mysql row conversion, err:{}", e);
+                        return Err(CommonErr(
+                            CommonErrCode::ConvertFailed,
+                            format!("mysql row conversion err:{e}"),
+                        ));
+                    }
+                };
+                parse_ob_vsn_from_sql(cluster_version);
+            }
+        }
+        Ok(())
+    }
+
+    // getTableEntryLocationFromRemote
```

Comment:
Such comment is meaningless, how about removing it or giving a more meaningful comment?

Reply:
Fixed